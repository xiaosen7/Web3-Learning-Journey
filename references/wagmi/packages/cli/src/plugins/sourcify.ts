import { Abi as AbiSchema } from 'abitype/zod'
import type { Address } from 'viem'
import type * as chain from 'viem/chains'
import { z } from 'zod'

import type { ContractConfig } from '../config.js'
import { fromZodError } from '../errors.js'
import type { Compute } from '../types.js'
import { fetch } from './fetch.js'

export type SourcifyConfig<chainId extends number> = {
  /**
   * Duration in milliseconds to cache ABIs.
   *
   * @default 1_800_000 // 30m in ms
   */
  cacheDuration?: number | undefined
  /**
   * Chain id to use for fetching ABI.
   *
   * If `address` is an object, `chainId` is used to select the address.
   *
   * See https://docs.sourcify.dev/docs/chains for supported chains.
   */
  chainId: chainId
  /**
   * Contracts to fetch ABIs for.
   */
  contracts: Compute<Omit<ContractConfig<ChainId, chainId>, 'abi'>>[]
}

const SourcifyResponse = z.object({
  compiler: z.object({
    version: z.string(),
  }),
  language: z.string(),
  output: z.object({
    abi: AbiSchema,
    devdoc: z.any(),
    userdoc: z.any(),
  }),
  settings: z.any(),
  sources: z.any(),
  version: z.number(),
})

/** Fetches contract ABIs from Sourcify. */
export function sourcify<chainId extends ChainId>(
  config: SourcifyConfig<chainId>,
) {
  const { cacheDuration, chainId, contracts: contracts_ } = config

  const contracts = contracts_.map((x) => ({
    ...x,
    address:
      typeof x.address === 'string' ? { [chainId]: x.address } : x.address,
  })) as Omit<ContractConfig, 'abi'>[]

  return fetch({
    cacheDuration,
    contracts,
    async parse({ response }) {
      if (response.status === 404)
        throw new Error('Contract not found in Sourcify repository.')

      const json = await response.json()
      const parsed = await SourcifyResponse.safeParseAsync(json)
      if (!parsed.success)
        throw fromZodError(parsed.error, { prefix: 'Invalid response' })

      if (parsed.data.output.abi)
        return parsed.data.output.abi as ContractConfig['abi']
      throw new Error('contract not found')
    },
    request({ address }) {
      if (!address) throw new Error('address is required')

      let contractAddress: Address | undefined
      if (typeof address === 'string') contractAddress = address
      else if (typeof address === 'object') contractAddress = address[chainId]

      if (!contractAddress)
        throw new Error(
          `No address found for chainId "${chainId}". Make sure chainId "${chainId}" is set as an address.`,
        )

      return {
        url: `https://repo.sourcify.dev/contracts/full_match/${chainId}/${contractAddress}/metadata.json`,
      }
    },
  })
}

// Supported chains
// https://docs.sourcify.dev/docs/chains
type ChainId =
  | typeof chain.mainnet.id
  | typeof chain.goerli.id
  | 11155111
  | typeof chain.arbitrumGoerli.id
  | typeof chain.arbitrum.id
  | 592
  | typeof chain.aurora.id
  | typeof chain.auroraTestnet.id
  | typeof chain.avalanche.id
  | typeof chain.avalancheFuji.id
  | 56
  | 97
  | 288
  | 28
  | 534
  | typeof chain.canto.id
  | typeof chain.celoAlfajores.id
  | 62320
  | typeof chain.celo.id
  | typeof chain.gnosisChiado.id
  | 103090
  | 53935
  | 335
  | 44
  | 43
  | 432204
  | 432201
  | 246
  | 73799
  | typeof chain.evmos.id
  | typeof chain.evmosTestnet.id
  | 122
  | 486217935
  | 192837465
  | 356256156
  | typeof chain.gnosis.id
  | 71402
  | 71401
  | 420420
  | 420666
  | 8217
  | 1001
  | 82
  | 83
  | 1287
  | 1284
  | 1285
  | 62621
  | 42262
  | 42261
  | 23295
  | 311752642
  | 4216137055
  | typeof chain.optimism.id
  | 28528
  | typeof chain.optimismGoerli.id
  | 300
  | 99
  | 77
  | 11297108109
  | 11297108099
  | typeof chain.polygon.id
  | typeof chain.polygonMumbai.id
  | 336
  | 57
  | 5700
  | 40
  | 41
  | 8
  | 106
  | 11111
  | 51
  | 7001
