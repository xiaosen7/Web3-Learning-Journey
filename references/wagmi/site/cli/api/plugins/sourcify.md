# sourcify

Plugin for fetching ABIs from [Sourcify](https://sourcify.dev/). Sourcify is a decentralized, open-source, smart contract verification and metadata repository.

## Import

```ts
import { sourcify } from '@wagmi/cli/plugins'
```

## Usage

```ts{2,6-13}
import { defineConfig } from '@wagmi/cli'
import { sourcify } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    sourcify({
      contracts: [
        {
          name: 'deposit',
          address: '0x00000000219ab540356cbb839cbe05303d7705fa',
        },
      ],
    }),
  ],
})
```

## Configuration

```ts
import { type SourcifyConfig } from '@wagmi/cli/plugins'
```

### cacheDuration

`number | undefined`

- Duration in milliseconds to cache ABIs.
- Defaults to `1_800_000` (30 minutes).

```ts
import { defineConfig } from '@wagmi/cli'
import { sourcify } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    sourcify({
      cacheDuration: 300_000, // [!code focus]
      chainId: 100,
      contracts: [
        {
          name: 'Deposit',
          address: '0x00000000219ab540356cbb839cbe05303d7705fa',
        },
      ],
    }),
  ],
})
```

### chainId

`number`

Chain ID to use for fetching ABI. If `address` is an object, `chainId` is used to select the address. See [Sourcify docs](https://docs.sourcify.dev/docs/chains) for supported chains.

```ts
import { defineConfig } from '@wagmi/cli'
import { sourcify } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    sourcify({
      chainId: 100, // [!code focus]
      contracts: [
        {
          name: 'Community',
          address: {
            100: '0xC4c622862a8F548997699bE24EA4bc504e5cA865',
            137: '0xC4c622862a8F548997699bE24EA4bc504e5cA865',
          },
        },
      ],
    }),
  ],
})
```

### contracts

`{ name: string; address?: Address | Record<number, Address> | undefined }[]`

Contracts to fetch ABIs for.

```ts
import { defineConfig } from '@wagmi/cli'
import { sourcify } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    sourcify({
      chainId: 100,
      contracts: [ // [!code focus]
        { // [!code focus]
          name: 'Deposit', // [!code focus]
          address: '0x00000000219ab540356cbb839cbe05303d7705fa', // [!code focus]
        }, // [!code focus]
      ], // [!code focus]
    }),
  ],
})
```
