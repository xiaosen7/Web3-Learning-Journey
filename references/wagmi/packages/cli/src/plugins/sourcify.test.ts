import { http, HttpResponse } from 'msw'
import { setupServer } from 'msw/node'
import { afterAll, afterEach, beforeAll, expect, test } from 'vitest'

import { depositAbi } from '../../test/constants.js'
import { sourcify } from './sourcify.js'

const baseUrl = 'https://repo.sourcify.dev/contracts/full_match'
const address = '0x00000000219ab540356cbb839cbe05303d7705fa'
const chainId = 1
const multichainAddress = '0xC4c622862a8F548997699bE24EA4bc504e5cA865'
const multichainIdGnosis = 100
const multichainIdPolygon = 137
const successJson = {
  compiler: { version: '0.6.11+commit.5ef660b1' },
  language: 'Solidity',
  output: {
    abi: depositAbi,
    devdoc: {},
    userdoc: {},
  },
  settings: {},
  sources: {},
  version: 1,
}

const handlers = [
  http.get(`${baseUrl}/${chainId}/${address}/metadata.json`, () =>
    HttpResponse.json(successJson),
  ),
  http.get(`${baseUrl}/${multichainIdGnosis}/${address}/metadata.json`, () =>
    HttpResponse.json({}, { status: 404 }),
  ),
  http.get(
    `${baseUrl}/${multichainIdGnosis}/${multichainAddress}/metadata.json`,
    () => HttpResponse.json(successJson),
  ),
  http.get(
    `${baseUrl}/${multichainIdPolygon}/${multichainAddress}/metadata.json`,
    () => HttpResponse.json(successJson),
  ),
]

const server = setupServer(...handlers)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('fetches ABI', () => {
  expect(
    sourcify({
      chainId: chainId,
      contracts: [{ name: 'DepositContract', address }],
    }).contracts?.(),
  ).resolves.toMatchSnapshot()
})

test('fetches ABI with multichain deployment', () => {
  expect(
    sourcify({
      chainId: 100,
      contracts: [
        {
          name: 'Community',
          address: { 100: multichainAddress, 137: multichainAddress },
        },
      ],
    }).contracts?.(),
  ).resolves.toMatchSnapshot()
})

test('fails to fetch for unverified contract', () => {
  expect(
    sourcify({
      chainId: 100,
      contracts: [{ name: 'DepositContract', address }],
    }).contracts?.(),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    '[Error: Contract not found in Sourcify repository.]',
  )
})

test('missing address for chainId', () => {
  expect(
    sourcify({
      chainId: 1,
      // @ts-expect-error `chainId` and `keyof typeof contracts[number].address` mismatch
      contracts: [{ name: 'DepositContract', address: { 10: address } }],
    }).contracts?.(),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `[Error: No address found for chainId "1". Make sure chainId "1" is set as an address.]`,
  )
})
