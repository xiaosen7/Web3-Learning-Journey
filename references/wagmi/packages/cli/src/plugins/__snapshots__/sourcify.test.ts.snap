// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetches ABI 1`] = `
[
  {
    "abi": [
      {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "inputs": [
          {
            "name": "pubkey",
            "type": "bytes",
          },
          {
            "name": "withdrawal_credentials",
            "type": "bytes",
          },
          {
            "name": "amount",
            "type": "bytes",
          },
          {
            "name": "signature",
            "type": "bytes",
          },
          {
            "name": "index",
            "type": "bytes",
          },
        ],
        "name": "DepositEvent",
        "type": "event",
      },
      {
        "inputs": [
          {
            "name": "pubkey",
            "type": "bytes",
          },
          {
            "name": "withdrawal_credentials",
            "type": "bytes",
          },
          {
            "name": "signature",
            "type": "bytes",
          },
          {
            "name": "deposit_data_root",
            "type": "bytes32",
          },
        ],
        "name": "deposit",
        "outputs": [],
        "stateMutability": "payable",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "get_deposit_count",
        "outputs": [
          {
            "type": "bytes",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "get_deposit_root",
        "outputs": [
          {
            "type": "bytes32",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "name": "interfaceId",
            "type": "bytes4",
          },
        ],
        "name": "supportsInterface",
        "outputs": [
          {
            "type": "bool",
          },
        ],
        "stateMutability": "pure",
        "type": "function",
      },
    ],
    "address": {
      "1": "0x00000000219ab540356cbb839cbe05303d7705fa",
    },
    "name": "DepositContract",
  },
]
`;

exports[`fetches ABI with multichain deployment 1`] = `
[
  {
    "abi": [
      {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "inputs": [
          {
            "name": "pubkey",
            "type": "bytes",
          },
          {
            "name": "withdrawal_credentials",
            "type": "bytes",
          },
          {
            "name": "amount",
            "type": "bytes",
          },
          {
            "name": "signature",
            "type": "bytes",
          },
          {
            "name": "index",
            "type": "bytes",
          },
        ],
        "name": "DepositEvent",
        "type": "event",
      },
      {
        "inputs": [
          {
            "name": "pubkey",
            "type": "bytes",
          },
          {
            "name": "withdrawal_credentials",
            "type": "bytes",
          },
          {
            "name": "signature",
            "type": "bytes",
          },
          {
            "name": "deposit_data_root",
            "type": "bytes32",
          },
        ],
        "name": "deposit",
        "outputs": [],
        "stateMutability": "payable",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "get_deposit_count",
        "outputs": [
          {
            "type": "bytes",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "get_deposit_root",
        "outputs": [
          {
            "type": "bytes32",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "name": "interfaceId",
            "type": "bytes4",
          },
        ],
        "name": "supportsInterface",
        "outputs": [
          {
            "type": "bool",
          },
        ],
        "stateMutability": "pure",
        "type": "function",
      },
    ],
    "address": {
      "100": "0xC4c622862a8F548997699bE24EA4bc504e5cA865",
      "137": "0xC4c622862a8F548997699bE24EA4bc504e5cA865",
    },
    "name": "Community",
  },
]
`;
